pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -310,6 +312,57 @@ func (vs *vSphereVMProvider) createOrUpdateVirtualMachine(
 	return chanErr, nil
 }
 
+// CleanupVirtualMachine cleans and sanitizes the vSphere VM before it
+// is unregistered from Supervisor. This is used when a VM custom
+// resource with
+// "vmoperator.vmware.com.protected/skip-delete-platform-resource"
+// annotation is deleted.
+func (vs *vSphereVMProvider) CleanupVirtualMachine(
+	ctx context.Context,
+	vm *vmopv1.VirtualMachine) error {
+
+	vmNamespacedName := vm.NamespacedName()
+
+	if _, ok := currentlyReconciling.Load(vmNamespacedName); ok {
+		// If the VM is already being reconciled in a goroutine then it cannot
+		// be cleaned up yet.  Return and requeue.
+		return providers.ErrReconcileInProgress
+	}
+
+	vmCtx := pkgctx.VirtualMachineContext{
+		Context: context.WithValue(ctx, vimtypes.ID{}, vs.getOpID(ctx, vm, "cleanupVM")),
+		Logger:  pkglog.FromContextOrDefault(ctx),
+		VM:      vm,
+	}
+
+	client, err := vs.getVcClient(vmCtx)
+	if err != nil {
+		return err
+	}
+
+	vmCtx.Context = pkgctx.WithRestClient(vmCtx.Context, client.RestClient())
```

Comment:
I don't think this is needed - maybe was before using Reconfigure to remove the tags